File: web-tunnel/shared/src/media-video-codec.ts
export const MEDIA_VIDEO_FRAME_WIDTH = 640;
export const MEDIA_VIDEO_FRAME_HEIGHT = 360;
export const MEDIA_VIDEO_GRID_COLS = 160;
export const MEDIA_VIDEO_GRID_ROWS = 90;
export const MEDIA_VIDEO_SYMBOL_BITS = 2;
export const MEDIA_VIDEO_PACKET_BYTES = (MEDIA_VIDEO_GRID_COLS * MEDIA_VIDEO_GRID_ROWS * MEDIA_VIDEO_SYMBOL_BITS) / 8;
export const MEDIA_VIDEO_PAYLOAD_BYTES = MEDIA_VIDEO_PACKET_BYTES - 17;

const MAGIC0 = 0x57; // W
const MAGIC1 = 0x54; // T
const MAGIC2 = 0x4d; // M
const VERSION = 2;
const HEADER_BYTES = 15;
const CRC_OFFSET = MEDIA_VIDEO_PACKET_BYTES - 2;
const NO_ACK = 0xffff;
const SYMBOL_VALUES = [24, 96, 160, 232] as const;

export const MediaVideoPacketKind = {
  Ack: 0,
  Data: 1,
} as const;
export type MediaVideoPacketKind = (typeof MediaVideoPacketKind)[keyof typeof MediaVideoPacketKind];

export interface MediaVideoPacket {
  kind: MediaVideoPacketKind;
  seq: number;
  ackSeq: number | null;
  msgId: number;
  fragIndex: number;
  fragCount: number;
  payload: Uint8Array;
}

export function encodeMediaVideoPacket(packet: MediaVideoPacket): Uint8Array {
  if (packet.payload.byteLength > MEDIA_VIDEO_PAYLOAD_BYTES) {
    throw new RangeError(`media-video payload too large (${packet.payload.byteLength} > ${MEDIA_VIDEO_PAYLOAD_BYTES})`);
  }
  if (!isU16(packet.seq) || !isU16(packet.msgId)) throw new RangeError('seq/msgId must fit in uint16');
  if (!isU8(packet.fragIndex) || !isU8(packet.fragCount)) throw new RangeError('fragment index/count must fit in uint8');
  const out = new Uint8Array(MEDIA_VIDEO_PACKET_BYTES);
  out[0] = MAGIC0;
  out[1] = MAGIC1;
  out[2] = MAGIC2;
  out[3] = VERSION;
  out[4] = packet.kind;
  writeU16(out, 5, packet.seq);
  writeU16(out, 7, packet.ackSeq == null ? NO_ACK : packet.ackSeq);
  writeU16(out, 9, packet.msgId);
  out[11] = packet.fragIndex;
  out[12] = packet.fragCount;
  writeU16(out, 13, packet.payload.byteLength);
  out.set(packet.payload, HEADER_BYTES);
  writeU16(out, CRC_OFFSET, crc16(out.subarray(0, CRC_OFFSET)));
  return out;
}

export function decodeMediaVideoPacket(bytes: Uint8Array): MediaVideoPacket | null {
  if (bytes.byteLength < MEDIA_VIDEO_PACKET_BYTES) return null;
  if (bytes[0] !== MAGIC0 || bytes[1] !== MAGIC1 || bytes[2] !== MAGIC2 || bytes[3] !== VERSION) return null;
  const expectedCrc = readU16(bytes, CRC_OFFSET);
  const actualCrc = crc16(bytes.subarray(0, CRC_OFFSET));
  if (expectedCrc !== actualCrc) return null;
  const payloadLen = readU16(bytes, 13);
  if (payloadLen > MEDIA_VIDEO_PAYLOAD_BYTES) return null;
  const kind = bytes[4]!;
  if (kind !== MediaVideoPacketKind.Ack && kind !== MediaVideoPacketKind.Data) return null;
  const ack = readU16(bytes, 7);
  return {
    kind,
    seq: readU16(bytes, 5),
    ackSeq: ack === NO_ACK ? null : ack,
    msgId: readU16(bytes, 9),
    fragIndex: bytes[11]!,
    fragCount: bytes[12]!,
    payload: bytes.slice(HEADER_BYTES, HEADER_BYTES + payloadLen),
  };
}

export function encodeMediaVideoPacketToI420(packetBytes: Uint8Array, width = MEDIA_VIDEO_FRAME_WIDTH, height = MEDIA_VIDEO_FRAME_HEIGHT): Uint8Array {
  if (packetBytes.byteLength !== MEDIA_VIDEO_PACKET_BYTES) {
    throw new RangeError(`media-video packet must be ${MEDIA_VIDEO_PACKET_BYTES} bytes`);
  }
  const chromaWidth = Math.trunc((width + 1) / 2);
  const chromaHeight = Math.trunc((height + 1) / 2);
  const yBytes = width * height;
  const out = new Uint8Array(yBytes + chromaWidth * chromaHeight * 2);
  fillGrid(out, width, height, packetBytes);
  out.fill(128, yBytes);
  return out;
}

export function decodeMediaVideoPacketFromI420(frameBytes: Uint8Array, width: number, height: number): Uint8Array {
  if (frameBytes.byteLength < width * height) {
    throw new RangeError('I420 frame is missing the luma plane');
  }
  const out = new Uint8Array(MEDIA_VIDEO_PACKET_BYTES);
  const symbols = MEDIA_VIDEO_PACKET_BYTES * (8 / MEDIA_VIDEO_SYMBOL_BITS);
  for (let symbolIndex = 0; symbolIndex < symbols; symbolIndex += 1) {
    const col = symbolIndex % MEDIA_VIDEO_GRID_COLS;
    const row = Math.floor(symbolIndex / MEDIA_VIDEO_GRID_COLS);
    const symbol = readCellSymbol(frameBytes, width, height, col, row);
    writeSymbol(out, symbolIndex, symbol);
  }
  return out;
}

export function encodeMediaVideoAckPayload(seqs: readonly number[]): Uint8Array {
  const maxSeqs = Math.floor(MEDIA_VIDEO_PAYLOAD_BYTES / 2);
  const count = Math.min(seqs.length, maxSeqs);
  const out = new Uint8Array(count * 2);
  for (let i = 0; i < count; i += 1) writeU16(out, i * 2, seqs[i]!);
  return out;
}

export function decodeMediaVideoAckPayload(payload: Uint8Array): number[] {
  const out: number[] = [];
  for (let i = 0; i + 1 < payload.byteLength; i += 2) out.push(readU16(payload, i));
  return out;
}

function fillGrid(yPlane: Uint8Array, width: number, height: number, packetBytes: Uint8Array): void {
  for (let row = 0; row < MEDIA_VIDEO_GRID_ROWS; row += 1) {
    const y0 = Math.floor((row * height) / MEDIA_VIDEO_GRID_ROWS);
    const y1 = Math.floor(((row + 1) * height) / MEDIA_VIDEO_GRID_ROWS);
    for (let col = 0; col < MEDIA_VIDEO_GRID_COLS; col += 1) {
      const symbolIndex = row * MEDIA_VIDEO_GRID_COLS + col;
      const value = SYMBOL_VALUES[readSymbol(packetBytes, symbolIndex)]!;
      const x0 = Math.floor((col * width) / MEDIA_VIDEO_GRID_COLS);
      const x1 = Math.floor(((col + 1) * width) / MEDIA_VIDEO_GRID_COLS);
      for (let y = y0; y < y1; y += 1) {
        yPlane.fill(value, y * width + x0, y * width + x1);
      }
    }
  }
}

function readCellSymbol(yPlane: Uint8Array, width: number, height: number, col: number, row: number): number {
  const x0 = Math.floor(((col + 0.25) * width) / MEDIA_VIDEO_GRID_COLS);
  const x1 = Math.max(x0 + 1, Math.floor(((col + 0.75) * width) / MEDIA_VIDEO_GRID_COLS));
  const y0 = Math.floor(((row + 0.25) * height) / MEDIA_VIDEO_GRID_ROWS);
  const y1 = Math.max(y0 + 1, Math.floor(((row + 0.75) * height) / MEDIA_VIDEO_GRID_ROWS));
  let sum = 0;
  let n = 0;
  for (let y = y0; y < y1; y += 1) {
    for (let x = x0; x < x1; x += 1) {
      sum += yPlane[y * width + x]!;
      n += 1;
    }
  }
  const avg = sum / Math.max(1, n);
  if (avg < 60) return 0;
  if (avg < 128) return 1;
  if (avg < 196) return 2;
  return 3;
}

function readSymbol(bytes: Uint8Array, symbolIndex: number): number {
  const byteIndex = Math.floor(symbolIndex / 4);
  const shift = 6 - (symbolIndex % 4) * 2;
  return (bytes[byteIndex]! >> shift) & 0x03;
}

function writeSymbol(bytes: Uint8Array, symbolIndex: number, symbol: number): void {
  const byteIndex = Math.floor(symbolIndex / 4);
  const shift = 6 - (symbolIndex % 4) * 2;
  bytes[byteIndex] = (bytes[byteIndex]! & ~(0x03 << shift)) | ((symbol & 0x03) << shift);
}

function crc16(bytes: Uint8Array): number {
  let crc = 0xffff;
  for (const b of bytes) {
    crc ^= b << 8;
    for (let i = 0; i < 8; i += 1) {
      crc = (crc & 0x8000) !== 0 ? ((crc << 1) ^ 0x1021) & 0xffff : (crc << 1) & 0xffff;
    }
  }
  return crc & 0xffff;
}

function isU16(value: number): boolean {
  return Number.isInteger(value) && value >= 0 && value <= 0xffff;
}

function isU8(value: number): boolean {
  return Number.isInteger(value) && value >= 0 && value <= 0xff;
}

function writeU16(out: Uint8Array, offset: number, value: number): void {
  out[offset] = (value >>> 8) & 0xff;
  out[offset + 1] = value & 0xff;
}

function readU16(bytes: Uint8Array, offset: number): number {
  return (bytes[offset]! << 8) | bytes[offset + 1]!;
}

File: web-tunnel/client-android/app/src/main/java/ai/webtunnel/mobile/MediaVideoCodec.kt
package ai.webtunnel.mobile

const val MEDIA_VIDEO_FRAME_WIDTH: Int = 640
const val MEDIA_VIDEO_FRAME_HEIGHT: Int = 360
const val MEDIA_VIDEO_GRID_COLS: Int = 160
const val MEDIA_VIDEO_GRID_ROWS: Int = 90
const val MEDIA_VIDEO_SYMBOL_BITS: Int = 2
const val MEDIA_VIDEO_PACKET_BYTES: Int = MEDIA_VIDEO_GRID_COLS * MEDIA_VIDEO_GRID_ROWS * MEDIA_VIDEO_SYMBOL_BITS / 8
const val MEDIA_VIDEO_PAYLOAD_BYTES: Int = MEDIA_VIDEO_PACKET_BYTES - 17

const val MEDIA_VIDEO_PACKET_KIND_ACK: Int = 0
const val MEDIA_VIDEO_PACKET_KIND_DATA: Int = 1

private const val MAGIC0 = 0x57
private const val MAGIC1 = 0x54
private const val MAGIC2 = 0x4d
private const val VERSION = 2
private const val HEADER_BYTES = 15
private const val CRC_OFFSET = MEDIA_VIDEO_PACKET_BYTES - 2
private const val NO_ACK = 0xffff
private val SYMBOL_VALUES = intArrayOf(24, 96, 160, 232)

data class MediaVideoPacket(
  val kind: Int,
  val seq: Int,
  val ackSeq: Int?,
  val msgId: Int,
  val fragIndex: Int,
  val fragCount: Int,
  val payload: ByteArray,
)

fun encodeMediaVideoPacket(packet: MediaVideoPacket): ByteArray {
  require(packet.payload.size <= MEDIA_VIDEO_PAYLOAD_BYTES) {
    "media-video payload too large (${packet.payload.size} > $MEDIA_VIDEO_PAYLOAD_BYTES)"
  }
  require(isU16(packet.seq) && isU16(packet.msgId)) { "seq/msgId must fit in uint16" }
  require(isU8(packet.fragIndex) && isU8(packet.fragCount)) { "fragment index/count must fit in uint8" }
  val out = ByteArray(MEDIA_VIDEO_PACKET_BYTES)
  out[0] = MAGIC0.toByte()
  out[1] = MAGIC1.toByte()
  out[2] = MAGIC2.toByte()
  out[3] = VERSION.toByte()
  out[4] = packet.kind.toByte()
  writeU16(out, 5, packet.seq)
  writeU16(out, 7, packet.ackSeq ?: NO_ACK)
  writeU16(out, 9, packet.msgId)
  out[11] = packet.fragIndex.toByte()
  out[12] = packet.fragCount.toByte()
  writeU16(out, 13, packet.payload.size)
  System.arraycopy(packet.payload, 0, out, HEADER_BYTES, packet.payload.size)
  writeU16(out, CRC_OFFSET, crc16(out, 0, CRC_OFFSET))
  return out
}

fun decodeMediaVideoPacket(bytes: ByteArray): MediaVideoPacket? {
  if (bytes.size < MEDIA_VIDEO_PACKET_BYTES) return null
  if ((bytes[0].toInt() and 0xff) != MAGIC0) return null
  if ((bytes[1].toInt() and 0xff) != MAGIC1) return null
  if ((bytes[2].toInt() and 0xff) != MAGIC2) return null
  if ((bytes[3].toInt() and 0xff) != VERSION) return null
  val expectedCrc = readU16(bytes, CRC_OFFSET)
  val actualCrc = crc16(bytes, 0, CRC_OFFSET)
  if (expectedCrc != actualCrc) return null
  val payloadLen = readU16(bytes, 13)
  if (payloadLen > MEDIA_VIDEO_PAYLOAD_BYTES) return null
  val kind = bytes[4].toInt() and 0xff
  if (kind != MEDIA_VIDEO_PACKET_KIND_ACK && kind != MEDIA_VIDEO_PACKET_KIND_DATA) return null
  val ack = readU16(bytes, 7)
  return MediaVideoPacket(
    kind = kind,
    seq = readU16(bytes, 5),
    ackSeq = if (ack == NO_ACK) null else ack,
    msgId = readU16(bytes, 9),
    fragIndex = bytes[11].toInt() and 0xff,
    fragCount = bytes[12].toInt() and 0xff,
    payload = bytes.copyOfRange(HEADER_BYTES, HEADER_BYTES + payloadLen),
  )
}

fun encodeMediaVideoPacketToI420(
  packetBytes: ByteArray,
  width: Int = MEDIA_VIDEO_FRAME_WIDTH,
  height: Int = MEDIA_VIDEO_FRAME_HEIGHT,
): ByteArray {
  require(packetBytes.size == MEDIA_VIDEO_PACKET_BYTES) {
    "media-video packet must be $MEDIA_VIDEO_PACKET_BYTES bytes"
  }
  val chromaWidth = (width + 1) / 2
  val chromaHeight = (height + 1) / 2
  val yBytes = width * height
  val out = ByteArray(yBytes + chromaWidth * chromaHeight * 2)
  fillGrid(out, width, height, packetBytes)
  out.fill(128.toByte(), yBytes, out.size)
  return out
}

fun decodeMediaVideoPacketFromI420(yPlane: ByteArray, width: Int, height: Int): ByteArray {
  require(yPlane.size >= width * height) { "I420 frame is missing the luma plane" }
  val out = ByteArray(MEDIA_VIDEO_PACKET_BYTES)
  val symbols = MEDIA_VIDEO_PACKET_BYTES * (8 / MEDIA_VIDEO_SYMBOL_BITS)
  for (symbolIndex in 0 until symbols) {
    val col = symbolIndex % MEDIA_VIDEO_GRID_COLS
    val row = symbolIndex / MEDIA_VIDEO_GRID_COLS
    writeSymbol(out, symbolIndex, readCellSymbol(yPlane, width, height, col, row))
  }
  return out
}

fun encodeMediaVideoAckPayload(seqs: List<Int>): ByteArray {
  val maxSeqs = MEDIA_VIDEO_PAYLOAD_BYTES / 2
  val count = minOf(seqs.size, maxSeqs)
  val out = ByteArray(count * 2)
  for (i in 0 until count) writeU16(out, i * 2, seqs[i])
  return out
}

fun decodeMediaVideoAckPayload(payload: ByteArray): List<Int> {
  val out = ArrayList<Int>(payload.size / 2)
  var offset = 0
  while (offset + 1 < payload.size) {
    out += readU16(payload, offset)
    offset += 2
  }
  return out
}

private fun fillGrid(yPlane: ByteArray, width: Int, height: Int, packetBytes: ByteArray) {
  for (row in 0 until MEDIA_VIDEO_GRID_ROWS) {
    val y0 = (row * height) / MEDIA_VIDEO_GRID_ROWS
    val y1 = ((row + 1) * height) / MEDIA_VIDEO_GRID_ROWS
    for (col in 0 until MEDIA_VIDEO_GRID_COLS) {
      val symbolIndex = row * MEDIA_VIDEO_GRID_COLS + col
      val value = SYMBOL_VALUES[readSymbol(packetBytes, symbolIndex)].toByte()
      val x0 = (col * width) / MEDIA_VIDEO_GRID_COLS
      val x1 = ((col + 1) * width) / MEDIA_VIDEO_GRID_COLS
      for (y in y0 until y1) {
        yPlane.fill(value, y * width + x0, y * width + x1)
      }
    }
  }
}

private fun readCellSymbol(yPlane: ByteArray, width: Int, height: Int, col: Int, row: Int): Int {
  val x0 = (((col + 0.25) * width) / MEDIA_VIDEO_GRID_COLS).toInt()
  val x1 = maxOf(x0 + 1, (((col + 0.75) * width) / MEDIA_VIDEO_GRID_COLS).toInt())
  val y0 = (((row + 0.25) * height) / MEDIA_VIDEO_GRID_ROWS).toInt()
  val y1 = maxOf(y0 + 1, (((row + 0.75) * height) / MEDIA_VIDEO_GRID_ROWS).toInt())
  var sum = 0
  var n = 0
  for (y in y0 until y1) {
    for (x in x0 until x1) {
      sum += yPlane[y * width + x].toInt() and 0xff
      n += 1
    }
  }
  val avg = sum / maxOf(1, n)
  return when {
    avg < 60 -> 0
    avg < 128 -> 1
    avg < 196 -> 2
    else -> 3
  }
}

private fun readSymbol(bytes: ByteArray, symbolIndex: Int): Int {
  val byteIndex = symbolIndex / 4
  val shift = 6 - (symbolIndex % 4) * 2
  return (bytes[byteIndex].toInt() ushr shift) and 0x03
}

private fun writeSymbol(bytes: ByteArray, symbolIndex: Int, symbol: Int) {
  val byteIndex = symbolIndex / 4
  val shift = 6 - (symbolIndex % 4) * 2
  val current = bytes[byteIndex].toInt() and 0xff
  bytes[byteIndex] = ((current and (0x03 shl shift).inv()) or ((symbol and 0x03) shl shift)).toByte()
}

private fun crc16(bytes: ByteArray, offset: Int, length: Int): Int {
  var crc = 0xffff
  for (i in offset until offset + length) {
    crc = crc xor ((bytes[i].toInt() and 0xff) shl 8)
    repeat(8) {
      crc = if ((crc and 0x8000) != 0) {
        ((crc shl 1) xor 0x1021) and 0xffff
      } else {
        (crc shl 1) and 0xffff
      }
    }
  }
  return crc and 0xffff
}

private fun isU16(value: Int): Boolean = value in 0..0xffff
private fun isU8(value: Int): Boolean = value in 0..0xff

private fun writeU16(out: ByteArray, offset: Int, value: Int) {
  out[offset] = ((value ushr 8) and 0xff).toByte()
  out[offset + 1] = (value and 0xff).toByte()
}

private fun readU16(bytes: ByteArray, offset: Int): Int =
  ((bytes[offset].toInt() and 0xff) shl 8) or (bytes[offset + 1].toInt() and 0xff)
